storage/innobase/trx/trx0trx.cc

```diff
@@ -918,6 +918,9 @@ trx_start_low(
 
 #ifdef WITH_WSREP
 	trx->xid.null();
+	/* Transactions may be reused after commit, where wsrep state flag
+	gets reset. Sync it again from the corresponding THD object. */
+	trx->wsrep = wsrep_on(trx->mysql_thd);
```

Comment:
We already do this at a higher level in `innobase_trx_init()`. Why does it need to be duplicated?

Reply:
InnoDB transactions may be reused (for example, when they are taken from the pool of transactions or during DDL operation when there may be multiple commit calls using the same `trx_t` object). In this case, wsrep state flag is reset at commit time and is never restored back (`innobase_trx_init()` doesn't seem to be called). The case with DDL was the issue that initially surfaced. The check added in `trx_t::commit_in_memory()` allowed to also catch another case with the pool of transactions.

Reply:
We don't set the flag, we sync it from THD object. If THD object has this flag set, then it should be in sync with `trx_t`.

The initial place for the fix was exactly `trx_start_for_ddl_low()`, but after adding the check in `trx_t::commit_in_memory()` it turned out that such a fix was insufficient. And again, we don't set the flag, we sync it from THD object. I think it's a rather safe change.

Reply:
In case `trx->mysql_thd` is `NULL` the flag will be `false`. So the code that doesn't have THD objects for `trx_t` assigned, will keep running with `trx->wsrep` set to `false` as before.

Reply:
Thank you for the thorough review. I will need to re-explore the issue with this added knowledge.

Reply:
Non-replicated changes should be run in a separate transaction, right?

Reply:
There is bug on ::store_view but that is for different MDEV

Reply:
In any case, I am not going to approve something that will duplicate the assignment of `trx_t::wsrep`. The natural place for it would be when the transaction is started.

Reply:
I updated the PR and made `trx->wsrep` to be set in a single place. Other changes were in the handlerton code to use `THD::wsrep` variable rather than `trx->wsrep`, and a fix to `Wsrep_schema::store_view()` to operate in it own transaction like other methods do.

Reply:
I think I understood what I did wrong. I accidentally reverted back `trx->wsrep = wsrep_on(thd);` assignment in `innobase_trx_init()` which we previously agreed to remove and that made all the tests pass. I will need to work on this PR more.

Reply:
Thank you. This is rather complex, but hopefully with the use of `rr record` and some data watchpoints or breakpoints you will be able to figure it out. I think that we really should try to fix the `trx->wsrep` mechanism, because it should incur much less overhead than the function calls.